feat: implement `profile`, `global` and `save` data types

UserIsntAvailable opened the pull request:
@dpinela I have mostly tested that the new types work correctly, I would appreciate if you can verify that `OnceSaveData` stuff is still working.

dpinela commented on src/Hooks.cs:
```diff
@@ -31,11 +51,48 @@ private static void Postfix(int saveSlot)
+    nameof(GameManager.SaveGame),
+    [typeof(int), typeof(System.Action<bool>), typeof(bool), typeof(AutoSaveName)]
+)]
+internal static class SaveDataSaveHook
```
Saving OnceSaveData here would require us to track whether or not we already saved it for that particular file; I can imagine some ways to do it but none as foolproof as simply saving it on save startup since that happens only once by definition.

BadMagic100 replied:
Yeah that makes sense. Also tbh even if we did make this change I can't imagine there'd be any harm to keep the cleanup at startup around for safety anyway

BadMagic100 commented:
> What happens if I have a global/profile data mod, I load into the mod, then I add a new field (let’s say it’s an int field with default value 3) and then I load in again? Is the value of the field 3 or default(int)?

3, that's how Newtonsoft deserialization works, it's not different from hk

Other question I'd have to double check the code to answer

UserIsntAvailable commented:
> I'm not sure how much more difficult it would be at that point to just fully integrate with restore points so that may also be worth considering, but I think either approach is fine for an initial release

I left out restore point functionality from this PR out, since it isn't required to add the new types. I wasn't intending to get this merged to release 1.0 right soon after, since I would prefer to get required/optional mods feature implemented before doing so (I'm currently working on it, for another PR).

I could implement restore points in this PR, but doing it on another one wouldn't change anything anyways, but that is up to you to decide.

> Also for future reference, it's easier to review changes if you push a new commit rather than amending and force pushing since it allows reviewing the diff; I can just squash merge at the end

I actually considered asking when I started doing PRs, what would you be prefer, but since you didn't point it out I imagine you were correct with it. GitHub actually lets you see the diff btw (Compare button), but for some reason you can only do it on the desktop view...

BadMagic100 commented:
> I left out restore point functionality from this PR out, since it isn't required to add the new types. I wasn't intending to get this merged to release 1.0 right soon after, since I would prefer to get required/optional mods feature implemented before doing so (I'm currently working on it, for another PR).

I'd like one of the mentioned flavors of restore point handling on this pr and I would like to get 1.0 out soon and deliver additional features iteratively afterwards, there are some people blocked on this and the release keeps stalling out. So we just need bare minimum to make sure we don't accidentally make a breaking change

flibber-hk commented:
Seeing as SaveDataMod now exists and OnceSaveDataMod has never been released, is there any reason to keep it around? My understanding is that any code which implements OnceSaveDataMod should be able to have no change in functionality if it is simply replaced with SaveDataMod with no further code changes, so it is simply introducing additional maintenance burden. @dpinela please correct me if I'm wrong

UserIsntAvailable commented:
> I'd like one of the mentioned flavors of restore point handling on this pr and I would like to get 1.0 out soon and deliver additional features iteratively afterwards, there are some people blocked on this and the release keeps stalling out. So we just need bare minimum to make sure we don't accidentally make a breaking change

On the interest of having this release as soon as possible, I implemented the discord suggestion, but instead of deleting the whole modded save data directory, I only deleted files from `ISaveDataMod`, otherwise every `IOnceSaveDataMod` would have been "broken" when restoring save points.

I also changed the directory where `ISaveDataMod` files are saved to implement the suggestion; keeping save data files alongside `OncePerSave` meant that I needed to move the directory somewhere else before clearing the `userX` directory (which would have make the implementation a lot more convoluted).

dpinela commented:
Restore points benefit from that immutability as well, since just as when saving, using one does not require touching the immutable save data at all.

flibber-hk commented:
Do you think it makes sense to make it by contract that a mod which implements both types of save data has OnceSaveData loaded before regular save data? I feel this is probably better based on what BadMagic and Pimpas said if there isn't a good reason otherwise

BadMagic100 commented:
Yeah I think that ordering would be sensible

dpinela commented:
For instance, it would conveniently let you store most placement information for IC in OnceSaveData and then just the state flags in mutable SaveData. Conversely, I can't think of any case where I'd want the opposite ordering.

BadMagic100 commented on src/Hooks.cs:
```diff
@@ -31,11 +51,48 @@ private static void Postfix(int saveSlot)
         foreach (var mod in mods)
         {
+            mod.LoadSaveData(saveSlot);
             mod.LoadOnceSaveData(saveSlot);
```
Per the comments on the comment page I think we should load once-data first, there's a pretty reasonable chance that the mutable data can have some dependence on the immutable data, but the other cannot be true by definition (since once-data has to be saved as the file starts and can't change after)
